### PR TITLE
Create interfaces for IonDotnet.Tree

### PR DESCRIPTION
**Pull request by avneet-toor-bq**
*Issue #, if available:* #17 #18 

*Description of changes:*
- Created interfaces for the various `IonXXX` types in `IonDotnet.Tree` package
- Constructors for various `IonXXX` types put in `ValueFactory.cs`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: IonDotnet/Tree/Impl/ValueFactory.cs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        {
+            return IonBool.NewNull();
+        }
+
+        public IIonBool NewBool(bool value)
+        {
+            return new IonBool(value);
+        }
+
+        public IIonClob NullClob()
```
Comment (tgregg) [inline IonDotnet/Tree/Impl/ValueFactory.cs]: `NullClob` -> `NewNullClob` for consistency.

File: IonDotnet/Tree/Impl/ValueFactory.cs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        {
+            return new IonNull();
+        }
+
+        public IIonSexp NewNullSexp()
+        {
+            return IonSexp.NewNull();
+        }
+
+        public IIonSexp NewSexp()
```
Comment (tgregg) [inline IonDotnet/Tree/Impl/ValueFactory.cs]: `NewSexp` -> `NewEmtpySexp` for consistency with `NewEmptyList`.

File: IonDotnet/Tree/Impl/ValueFactory.cs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        {
+            return new IonSexp();
+        }
+
+        public IIonString NewString(string value)
```
Comment (tgregg) [inline IonDotnet/Tree/Impl/ValueFactory.cs]: Missing `NewNullString`.

File: IonDotnet/Tree/Impl/ValueFactory.cs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        {
+            return IonSymbol.NewNull();
+        }
+
+        public IIonSymbol NewSymbol(SymbolToken symbolToken)
+        {
+            return new IonSymbol(symbolToken);
+        }
+
+        public IIonSymbol NewSymbol(string text, int sid)
```
Comment (tgregg) [inline IonDotnet/Tree/Impl/ValueFactory.cs]: Let's not expose `sid`, which would be the local sid. Just make this `public IIonSymbol NewSymbol(string text)` and pass UNKNOWN_SID under the hood. If users want to construct symbols with unknown text they can use `public IIonSymbol NewSymbol(SymbolToken symbolToken)`.

File: IonDotnet/Tree/IValueFactory.cs [hunk 5]
```diff
@@ -0,0 +1,195 @@
+﻿using System;
+using System.Numerics;
+
+namespace IonDotnet.Tree
+{
+    /// <summary>
+    /// The factory for all {@link IonValue}s.
+    /// WARNING: This interface should not be implemented or extended by
```
Comment (tgregg) [inline IonDotnet/Tree/IValueFactory.cs]: We warned against implementing most of our public interfaces in ion-java and there's now fairly general agreement that that was a mistake. Let's just remove this guidance.

**Comment by tgregg** [comment]
Fixes #17 
Fixes #18

File: IonDotnet/Tree/Impl/ValueFactory.cs (hunk 1 above)
Comment (avneet-toor-bq) [reply IonDotnet/Tree/Impl/ValueFactory.cs]: Done.

File: IonDotnet/Tree/Impl/ValueFactory.cs (hunk 2 above)
Comment (avneet-toor-bq) [reply IonDotnet/Tree/Impl/ValueFactory.cs]: Done.

File: IonDotnet/Tree/Impl/ValueFactory.cs (hunk 3 above)
Comment (avneet-toor-bq) [reply IonDotnet/Tree/Impl/ValueFactory.cs]: Added. Also added a `NewNull()` method to `IonString.cs` to accommodate for this.

File: IonDotnet/Tree/Impl/ValueFactory.cs (hunk 4 above)
Comment (avneet-toor-bq) [reply IonDotnet/Tree/Impl/ValueFactory.cs]: Done. Under the hood, we set the SID to `UnknownSid` in `IonSymbol`'s constructor.

File: IonDotnet/Tree/IValueFactory.cs (hunk 5 above)
Comment (avneet-toor-bq) [reply IonDotnet/Tree/IValueFactory.cs]: Removed.

**Comment by avneet-toor-bq** [comment]
> Please merge in the latest changes from `master` to resolve conflicts and update this branch. Then I'll squash and merge this to `master`.

Done.

**Pull request merged**